dqvm/src/permutation.lisp

```diff
@@ -43,119 +88,114 @@ DQVM2> (make-permutation '((2 . 1) (1 . 0)))
-             (rest transposition)
-             item)
-           item)))
-
-(defun-inlinable apply-inverse-permutation (permutation item)
-  "Apply PERMUTATION⁻¹ to ITEM."
-  (apply-permutation (inverse-permutation permutation) item))
+(defmethod identity-permutation-p ((permutation permutation-general))
+  (null (permutation-transpositions permutation)))
 
 (defun compose-permutations (&rest permutations)
   "Return a new permutation that is the composition of PERMUTATIONS.
 
 If PERMUTATIONS is the list π₁, π₂, ..., πₛ, then the result is the composition π₁ ∘ π₂ ∘ ... ∘ πₛ. In other words, the composition starts from right to left as in standard mathematical notation."
-  (let (transpositions)
-
-    (let (domain)
-      ;; Aggregate the domain of the composed permutation to get a list of
-      ;; all possible relevant inputs.
-      (loop :for permutation :in permutations :when permutation :do
-        (loop :for transposition :of-type transposition :in (permutation-transpositions permutation) :do
-          (let ((a (first transposition)))
-            (declare (type alexandria:non-negative-fixnum a))
-            (pushnew a domain))))
-
-      ;; Now map each domain element to obtain transpositions.
-      (loop :with codomain := (coerce domain 'vector)
-            :for permutation :in (nreverse permutations) :when permutation :do
-              (loop :for i :from 0 :for b :across codomain :do
-                (setf (aref codomain i)
-                      (apply-permutation permutation (aref codomain i))))
-            :finally
-               (loop :for a :of-type alexandria:non-negative-fixnum :in domain
-                     :for b :of-type alexandria:non-negative-fixnum :across codomain
-                     :unless (= a b) :do
-                       (pushnew (cons a b) transpositions :test #'equal))))
+  (let ((transpositions nil)
+        (domain nil))
+
+    ;; Aggregate the domain of the composed permutation to get a list of
+    ;; all possible relevant inputs.
+    (loop :for permutation :in permutations :when permutation :do
+      (loop :for transposition :of-type transposition :in (permutation-transpositions permutation) :do
+        (let ((a (first transposition)))
+          (declare (type qubit-index a))
+          (pushnew a domain))))
+
+    ;; Now map each domain element to obtain transpositions.
+    (loop :with codomain := (coerce domain 'vector)
+          :for permutation :in (nreverse permutations) :when permutation :do
+            (loop :for i :from 0 :for b :across codomain :do
+              (setf (aref codomain i)
+                    (apply-permutation permutation (aref codomain i))))
+          :finally
+             (loop :for a :of-type qubit-index :in domain
+                   :for b :of-type qubit-index :across codomain
+                   :unless (= a b) :do
+                     (pushnew (cons a b) transpositions :test #'equal)))
```

Comment:
I would write a `transposition=` function and use it here, even if it's as simple as
```
(defun transposition= (a b)
  (and (= (car a) (car b)) (= (cdr a) (cdr b))))
```